### PR TITLE
Add support for tendermint 0.25

### LINKED ISSUE #295: Upgrade tendermint client to support 0.24 (and 0.23?)
We only support 0.20 and 0.21 for now.

0.22 seems like it was a bit of a refactoring series, and no need to add support to it now.

0.24 has quite a few fixes and we should plan to upgrade the backend in the near future, so we will need client support. If we can also get 0.23 without much extra work, why not?

### PR DESCRIPTION
**Pull request by ethanfrey**
Closes #295 

- [x] Update request formatting (integers as strings)
- [ ] Update response formatting (quite a bit)
- [ ] Test compatibility with basic tendermint rpc kvstore
- [ ] Test compatibility in bns

### CODE REVIEW

**Comment by ethanfrey** [comment]
This is ready for review.
Tendermint 0.25 tested in the CI.

I will make a separate PR for weave 0.9 support, maybe yet another for testing multiple tendermint versions at once (eg. 0.21 and 0.25)

File: packages/iov-tendermint-rpc/src/client.spec.ts [hunk 1]
```diff
@@ -208,197 +399,13 @@ describe("Client", () => {
     expect(info3).toBeTruthy();
   });
 
-  describe("With HttpClient", () => {
-    kvTestSuite(() => new HttpClient(tendermintUrl));
+  describe("With HttpClient: v0-20", () => {
```
Comment (webmaster128) [inline packages/iov-tendermint-rpc/src/client.spec.ts]: Remove (wrong) version from description.

File: packages/iov-tendermint-rpc/src/client.spec.ts [hunk 2]
```diff
@@ -208,197 +399,13 @@ describe("Client", () => {
...
-
-        const transaction1 = buildKvTx(randomId(), randomId());
-        const transaction2 = buildKvTx(randomId(), randomId());
-
-        await client.broadcastTxCommit({ tx: transaction1 });
-        await client.broadcastTxCommit({ tx: transaction2 });
-      })().catch(fail);
-    });
+    const factory = () => new WebsocketClient(tendermintUrl, onError);
+    websocketTestSuite(factory, v0_25);
```
Comment (webmaster128) [inline packages/iov-tendermint-rpc/src/client.spec.ts]: This adds some strange nesting that is hard to follow. We can keep it flat as before:

```ts
const factory = () => new WebsocketClient(tendermintUrl, onError);
kvTestSuite(factory, v0_25);
websocketTestSuite(factory, v0_25);
```

File: packages/iov-tendermint-rpc/src/client.spec.ts [hunk 3]
```diff
@@ -188,6 +188,197 @@ function kvTestSuite(rpcFactory: () => RpcClient): void {
   });
 }
 
+function websocketTestSuite(rpcFactory: () => RpcClient, adaptor: Adaptor): void {
+  // don't print out WebSocket errors if marked pending
```
Comment (webmaster128) [inline packages/iov-tendermint-rpc/src/client.spec.ts]: Comment belongs to `const onError = skipTests() ? () => 0 : console.log;`

File: packages/iov-tendermint-rpc/src/client.spec.ts [hunk 4]
```diff
@@ -188,6 +188,197 @@ function kvTestSuite(rpcFactory: () => RpcClient): void {
...
+  kvTestSuite(rpcFactory, adaptor);
+
+  it("can subscribe to block header events", done => {
+    pendingWithoutTendermint();
+
+    // seems that tendermint just guarantees within the last second for timestamp
+    const testStart = ReadonlyDate.now() - 1000;
```
Comment (webmaster128) [inline packages/iov-tendermint-rpc/src/client.spec.ts]: Could you try `expect(event.time.getTime()).toBeGreaterThanOrEqual(testStart);` instead of this unintuitive substraction?

File: packages/iov-tendermint-rpc/src/client.spec.ts [hunk 5]
```diff
@@ -188,6 +188,197 @@ function kvTestSuite(rpcFactory: () => RpcClient): void {
...
+        complete: () => fail("Stream must not close just because we don't listen anymore"),
+      });
+    })().catch(fail);
+  });
+
+  it("can subscribe to block events", done => {
+    pendingWithoutTendermint();
+
+    // seems that tendermint just guarantees within the last second for timestamp
+    const testStart = ReadonlyDate.now() - 1000;
```
Comment (webmaster128) [inline packages/iov-tendermint-rpc/src/client.spec.ts]: same here

File: packages/iov-tendermint-rpc/src/client.spec.ts (hunk 4 above)
Comment (webmaster128) [reply packages/iov-tendermint-rpc/src/client.spec.ts]: I see. I am fine with leaving the logic as it is. But moving the `- 1000` from `testStart` into the matcher makes the situation much clearer

Comment (ethanfrey) [reply packages/iov-tendermint-rpc/src/client.spec.ts]: Nice

File: packages/iov-tendermint-rpc/src/encodings.ts [hunk 6]
```diff
@@ -44,8 +44,21 @@ export function may<T, U>(transform: (val: T) => U, value: T | null | undefined)
...
+export class Integer {
+  public static parse(str: IntegerString): number {
+    return Int53.fromString(str).toNumber();
+  }
+
+  public static encode(num: number): IntegerString {
+    return new Int53(num).toString() as IntegerString;
+  }
+
+  public static ensure(n: number): number {
```
Comment (webmaster128) [inline packages/iov-tendermint-rpc/src/encodings.ts]: If you don't trust the runtime type to be the type system type, this shout change to `n: unknown`.

Also since this is called Integer.ensure, we need at least a [Number.isInteger](https://developer.mozilla.org/de/docs/Web/JavaScript/Reference/Global_Objects/Number/isInteger) test

Comment (ethanfrey) [reply packages/iov-tendermint-rpc/src/encodings.ts]: Yeah, the docs say it is a string, but in tests it was an integer.
I wanted to make sure it was there

File: packages/iov-tendermint-rpc/src/encodings.ts [hunk 7]
```diff
@@ -53,11 +53,14 @@ export class Integer {
...
 
-  public static ensure(n: number): number {
-    if (typeof n !== "number") {
-      throw new Error(`${n} is not a number`);
+  public static ensure(num: unknown): number {
+    if (typeof num !== "number" || !Number.isInteger(num)) {
+      throw new Error(`${num} is not a number`);
     }
-    return n;
+    if (!Number.isInteger(num)) {
```
[the remark this answers is not in the document]
Comment (ethanfrey) [reply packages/iov-tendermint-rpc/src/encodings.ts]: Not any more. Amended that commit.
